### DISCUSSION TITLE
Ui improvments

### DISCUSSION THREAD

**Discussion by Cwpute** [plateaukao/einkbro#236] [Ideas]
Hi everyone!
I'm glad to see some improvments have been made to the UI since update 10.0 but i think there is a lot more that can be done. Here is a big feedback and suggestion text i've been preparing since before 10.0, i've updated it since then but some new things might have slipped by me. Let's get into it!

> As an introduction and full disclaimer: i am using Einkbro on a regular smartphone which isn't mounted with an eink screen. I'd love to have one someday but it's just out of my financial reach. Also, i have dipped my toes into coding before but am not capable of producing anything that works in a reasonable amount of time. Here i can only provide feedback, ideas, and general support. Sorry for that.

Now to the subject that brings us here…

After some time with Einkbro (a few months), i can say i really like this browser and its options, but have constantly been impeded by its UI. I understand it is as it is for two main reasons:
1. **The less eink screens pixels are updated, the better**, which justifies having so many options visible by default, having less content hidden behind menus etc. This is a specificity i didn't have in mind at first.
2. **This app is optimized for tablets, not for smartphones**,  which again, allows for more visible buttons and options, why not use that available space ? The are a lot of eink tablets out there, but eink smartphones are just around the corner, support for them should probably be emphasized as well.

Here are issues i've encoutered which i'll try to adress one by one. I encourage you comment and iterate upon these in your answers:

1. DEFAULT TOOLBAR IS UNFIT FOR SMARTPHONES

![Screenshot_20230409-153039_1](https://user-images.githubusercontent.com/80604512/233775752-957fee1d-fc22-4784-bc60-1b04c51ff22b.png)

On a fresh install, the toolbar is so cluttered that it can't be used as is. Some default buttons have overlapping purposes coupled with unintuitive icons. On my 720x1520 screen, the adress bar is even cut-off to the side and useless.

For *smartphones*, this is what i'd advise the default toolbar be:
![Screenshot_20230409-161627_1](https://user-images.githubusercontent.com/80604512/233775799-7be02ef8-f89b-4e34-8e62-ce4a18654c18.png)

Let me explain eacf choice:
* No adress bar: you probably already know which website you're on so you don't need that info. If you still need to get the adress of that website, you can use the "Copy link" tool. Alternatively, the behavior for a long press on a tab in the Tab preview window could be changed to Copy link, as it currently closes the tab, which can already be done with a single click on the cross next to the tab's name. If you want the website's name, its written in the Tab preview window. If you want to input an adress or search for a keyword, the New tab button (+) serves that purpose already.
* Tab preview: as stated previously, it serves many purposes and needs to stay. its multi-purpose can very much be improved upon even, as suggested right above.
* New tab: acces to both bookmarks, history, and adress/search bar so it needs to stay too
* Reload/Stop loading: i think it may be useful to eink screens ? but i dont know its use cases, please share if you do use it often
* Close tab: prevents the user from opening the Tab preview for a common action. but perhaps it can be hidden as well.
* No back button: this is probably controversial, as i don't know what is the current standard for smartphones. Do they all still have the 3 default android buttons at the bottom of the screen, like i do ? If they do, there is no need for a duplicate back button on the app toolbar.
* No touch settings: this button links to navigation options the user would probably need on their first use, or later on once or twice, but certainly not constantly. If touch options are essential to set up on the app first use, the user should then be prompted with a welcome screen on that occasion, presenting them with touch options among other prime info and parameters. But it shouldn't appear on the toolbar. I may be wrong, but again, please do share your insight if you use that button often enough.
* Reader mode: always useful for eink screens to reduce load time and improve readability
* Font size: a useful feature for eink screens now has its place on the toolbar

As a refererence, this is what my own toolbar looks like. Again, my screen is not eink so i can afford navigating through some aditional windows and menus:
![Screenshot_20230409-161903_1](https://user-images.githubusercontent.com/80604512/233776044-14d7f1cc-3aad-45b6-83a0-550b1b386c93.png)
As you can see it's got a lot less buttons, but i can afford uponening *some* menus to find options and actions.

2. TOOLBAR TOOLS AND FOLDED MENU TOOLS SHOULD BE REORGANIZED AND SYNERGIZED

Here is what the folded menu looked like before 10.0:
![Screenshot_20230406-110755_1](https://user-images.githubusercontent.com/80604512/233776230-7d00fcad-8932-4df5-a897-6be37036ece2.png)

Here is what the folded menu looks like after 10.0:
![Screenshot_20230422-113816_1](https://user-images.githubusercontent.com/80604512/233776365-1905b88f-f0d1-479d-9e9d-0f8c844fa637.png)

Here is a suggested organisation for the folded menu [this was partially adressed in 10.0, so this section can be outdated] :

```
1st row: browser tools
Home, New bookmark, Save as, Open, Downloads folder, Close tab
2nd row: page tools
Translate, Read contents, Search here, Send to (app), Send/Receive Data
3d row: readability
Reader view, Desktop view, Vertical/horizontal page, Split screen, Font options, White background
4th row:
Advanced options, Quit
```

Below, related remarks and suggestions regarding the toolbar:
* Send Data and Receive data could be merged into one button. Also, isn't this covered by the android Send to/Share with feature already ?
* Quick toggle: if it's possible to toggle (almost) everything from there, why is there an in-depth option menu available ? how is this one quicker ? why can't these toggles be directly integrated as selectable toolbar tools ?
* Condense all font options behind a single Font button and in a sinqle Font menu pop-uo, which is almost already the case, except for bold, and custom fonts.
* Define As Home is so seldom used, having it only appear in the options menu should be enough. Alternatively you could enable it with a long press on the Home button.
* Home url is called Favourite in toolbar. Let's unify all these names under Home.
* Open With acts just the same as Share With. is it necessary to have two buttwns for the same behavior ? maybe condense the two into one button.
* Save As should include all available options in a single menu archive(mht), pdf, epub, shortcut on desktop. If the user do have a preference of extensiwn, let them handle with options in the Advanced options.
* Bookmarks tool actions should be reproductible in the bookmarks list that appears above the search bar when it is selected.
* Toolbar settings is accessible through the options menu already. Then it's not necessary to show them in the folding menu too.

**Ultimately, Toolbar tools, quick toggles, and menu actions/tools shouldn't be different from each other**: every option, action or toggle should be available for pinnig on the toolbar, pinned on the folded menu, or in a "More" as the user sees fit. This probably needs quite a lot of refactoring but would greatly eases the personalization of the browser and ultimately its efficiency.
Also, as you add more and more tools  (ChatGPT for example) this approach will help you and other users handle the tools at their dirposal as they see fit.

3. REORGANIZE ADVANCED OPTIONS MENU

Here is below a suggested reorganization for the Advanced options menu:

```
APPEARANCE
>vertical view
>tab bar visibility (hidden, shown)
>reserved height
>toolbar position
>trim adress http
>_separator_
>font options
>>default font
>>size
>>bold or not
>force desktop/mobile view
>preferred translation
>force reader mode
>>force vertical reader mode
>dark mode
-
NAVIGATION [or INTERACTION, or both]
>home url
>new tab behavior
>back button behavior
>keep screen awake
>confirm closing tabs
>Search
>>[all options from Search]
>Toolbar
>>visible/hidden tools [menu similar to Toolbar settings to set toolbar tools]
>>toolbar position (top, bottom)
>>hide toolbar when scrolling
>>>back button behavior (back, show toolbar)
>Media
>>force video in fullscreen
>>small windowed video
>>background play
>>content zooming
>volume keys for page scrolling
>use up-down commands
>Gestures
>>[all options from Touch settings tool]
>VI key bindings
-
TOOLS
>[suggested] "Save as" preferred format
>printed pdf paper size
>ChatGPT
>>[all options for ChatGPT]
-
SECURITY & PRIVACY
>[everything from Start Settings]
>save active tabs when closing app
>clear recently used bookmarks
>prune tracking querry param
>Synchronize bookmarks
>Save parameters (app data)
>>[all options from Saves]
>Adblock
>>[all adblock related options]
>[everything from Data Settings]
>Custom User Agent
>>[all options from Custom user agent]
-
DEBUG TOOLS [those can even be hidden in the ABOUT menu as some apps do]
>ssl error
>debug webview
-
ABOUT EINKBRO
```

Below, related remarks and suggestions about the Advanced options:
* Right now, toggleable options appear as ticked/unticked boxes in the quick toggle menu, or lone ticks without boxes in the advanced options menu. Also in the options menu, the ticks can appear above text. Let's unify their representation with either boxes, or with slide twggles like you can see in every other app now.
![Screenshot_20230422-110531_1](https://user-images.githubusercontent.com/80604512/233776915-0bf66197-785c-4912-a516-2610d218573c.png)
* Not every option or toggle needs an icon: it diverts the eye and information is more unclear as a result. One icon for each category of options is enough. In my suggested reworked Advanced options menu for example, you could have icon each main section, and in each section if several options are related, one for the first of the group.

THAT IS ALL 😱 yeah sorry that's a lot to take in but there is a lot tw say too. Hope this all gives you a good insight and provides imoortant feedback.

**Comment by gety9** [comment]
@Cwpute @plateaukao 

Interesting and well thought out suggestions if one aims to please mobile (small screen users), but with main user audience of eink bro being eink device owners (meaning tablet owners, cause that's by far the majority of eink devices) i think above suggestions will hurt. 

I use Eink bro on tablet and mobile. I agree that on mobile UX is a bit suboptimal, but it's more than compensated by it's table UX (in my opinion tablet UX (layout, sizes, features) is absolutely best from all the options (Chrome, Brave, Vivavldi, etc) - simply by far best)

-------
I am only 1 user, but from my opinion Daniel is doing amazing job, he created amazing project, Eink bro is great. Please don't change anything (design wise) !!!!! :)

Cpwut, one more time your suggestions are great, but imho, they are mainly great for mobile users (and we have lots of options for them, but very few for tablet users like most of us eink device owners).

**Comment by gety9** [comment]
@Cwpute, regarding toolbar

"For smartphones, this is what i'd advise the default toolbar be:
....
Let me explain eacf choice:
..."

i think you can achieve most of toobar things you are asking for right now in settings:

Go to [menu] > [Toolbar settings] and deselect the ones you don't need, and reorange the ones you need. With this you will have as few elements/buttons in tool bar as you need.

**Comment by Cwpute** [comment]
I do aknowledge that einkbro was designed for tablets and eink devices. In fact, i start my suggestion by clearly stating it.

This discussion has gone 8 months without being seemingly acknowledged by the maintainer, or anyone else for that matter. In that time, a lot of new features have been added, which would require an even bigger rework of my original suggestion. I also switched to another mobile focused browser in the meantime.
If mobile browsing ever gets a higher priority in this project, maybe this discussion will become relevant again, but until then, i won't be actively looking for answers or progress on the matter.

Le 11 janvier 2024 01:40:45 GMT+01:00, gety9 ***@***.***> a écrit :
